### ISSUE TITLE
Feature-005b: Telemetry Annotation System with Anti-Contamination

### ISSUE THREAD

**Issue by clayroach** [clayroach/otel-ai#113]
# Feature-005b: Telemetry Annotation System with Anti-Contamination

## Status: Design Phase 🔄

**Feature ID**: FEAT-005b  
**Parent Feature**: Feature-005 (Diagnostics UI Fine Tuning)  
**Related**: 
- Feature-005a (OTLP Capture & Replay) - Works with capture sessions
- Feature-012 (Model Fine-Tuning Pipeline) - Provides training data
- ADR-016 (LLM-Guided SRE Diagnostics) - Captures investigation journey

**Priority**: HIGH  
**Target Release**: Phase 3 Test Infrastructure  
**Created**: 2025-09-24

## Executive Summary

Implement a separate annotation system that maintains telemetry data immutability while capturing rich diagnostic metadata, test scenarios, and investigation insights. This system ensures clean separation between core telemetry and annotations to prevent model training contamination while enabling comprehensive diagnostic capabilities.

## Core Principles

1. **Immutable Traces**: Trace data is never modified after storage
2. **Separate Annotations Table**: All metadata stored separately, referencing traces
3. **Prefix-Based Namespacing**: Clear convention for annotation types
4. **Anti-Contamination**: Strict separation for clean model training
5. **Full Audit Trail**: Every annotation tracked with author and timestamp

### Database Schema

```sql
-- Traces remain pure and immutable
CREATE TABLE traces (
    trace_id String,
    span_id String,
    parent_span_id String,
    service_name LowCardinality(String),
    operation_name LowCardinality(String),
    start_time DateTime64(9),
    end_time DateTime64(9),
    duration_ns UInt64,
    status_code LowCardinality(String),
    attributes Map(String, String),  -- Original attributes only
    -- ... other fields
) ENGINE = MergeTree()
ORDER BY (service_name, start_time, trace_id);

-- Separate annotations table
CREATE TABLE annotations (
    annotation_id String DEFAULT generateUUIDv4(),
    
    -- Target references (at least one required)
    span_id Nullable(String),
    trace_id Nullable(String),
    time_range_start Nullable(DateTime64(9)),
    time_range_end Nullable(DateTime64(9)),
    service_name Nullable(String),
    
    -- Annotation content with prefix convention
    annotation_type LowCardinality(String), -- 'test', 'diag', 'human', 'llm', 'meta', 'train'
    annotation_key String,
    annotation_value String,  -- JSON-encoded
    confidence Nullable(Float32),
    
    -- Audit trail
    created_at DateTime64(9) DEFAULT now64(9),
    created_by String,  -- 'system:component', 'human:user-id', 'llm:model-name'
    session_id Nullable(String),
    
    -- Relationships
    parent_annotation_id Nullable(String),
    related_annotation_ids Array(String),
    
    INDEX idx_span_id span_id TYPE bloom_filter(0.01) GRANULARITY 1,
    INDEX idx_trace_id trace_id TYPE bloom_filter(0.01) GRANULARITY 1,
    INDEX idx_session session_id TYPE bloom_filter(0.01) GRANULARITY 1,
    INDEX idx_type annotation_type TYPE set(10) GRANULARITY 1
) ENGINE = MergeTree()
ORDER BY (annotation_type, created_at, annotation_id);
```

## Annotation Type Prefixes

| Prefix | Type | Purpose | When Added | Appears in Production |
|--------|------|---------|------------|----------------------|
| `test.` | Test Scenario | Feature flags, injected issues | Testing only | **NEVER** |
| `diag.` | Diagnostic | Hypotheses, evidence, conclusions | Investigation | Yes |
| `human.` | Human Insight | SRE observations and validations | Manual | Yes |
| `llm.` | AI Analysis | Model predictions and patterns | AI analysis | Yes |
| `meta.` | Metadata | Session info, capture context | Various | Yes |
| `train.` | Training | Labels for model training | Data prep | **NEVER** |

### Annotation Service

```typescript
export interface AnnotationService extends Context.Tag<"AnnotationService", {
  readonly addAnnotation: (
    annotation: Annotation
  ) => Effect.Effect<string, AnnotationError | ValidationError, never>
  
  readonly getAnnotationsForSpan: (
    spanId: string
  ) => Effect.Effect<readonly Annotation[], AnnotationError, never>
  
  readonly getAnnotationsForTrace: (
    traceId: string
  ) => Effect.Effect<readonly Annotation[], AnnotationError, never>
  
  readonly getAnnotationsForTimeRange: (
    range: TimeRange
  ) => Effect.Effect<readonly Annotation[], AnnotationError, never>
  
  readonly getAnnotationsByType: (
    type: AnnotationType,
    sessionId?: string
  ) => Effect.Effect<readonly Annotation[], AnnotationError, never>
}> {}
```

### Anti-Contamination Service

```typescript
export interface AnnotationSanitizer extends Context.Tag<"AnnotationSanitizer", {
  readonly sanitizeForTraining: (
    traces: readonly Trace[]
  ) => Effect.Effect<readonly Trace[], never, never>
  
  readonly validateNoContamination: (
    traces: readonly Trace[]
  ) => Effect.Effect<void, ValidationError, never>
  
  readonly detectProhibitedPrefixes: (
    traces: readonly Trace[],
    environment: 'production' | 'training' | 'test'
  ) => Effect.Effect<readonly Alert[], never, never>
}> {}
```

### Feature Flag Capture (test.*)

When feature flags are activated in the OTel Demo:
1. Flag controller records state change
2. Annotation service creates time-range annotation
3. `test.feature_flag` annotation links to affected time window
4. Never appears in production data

### Investigation Journey (diag.*, human.*, llm.*)

During incident investigation (ADR-016):
1. LLM forms hypothesis → `diag.hypothesis` annotation
2. Human adds observation → `human.observation` annotation  
3. AI detects pattern → `llm.pattern_detected` annotation
4. All annotations reference specific traces/spans
5. Full investigation timeline preserved

### Training Data Pipeline

```typescript
// Clean separation for training
const trainingData = {
  // Pure traces without annotations
  input: await storage.queryTraces(filter),
  
  // Annotations for validation only
  groundTruth: await annotationService.getAnnotationsByType('test'),
  
  // Human validations for quality scoring
  humanValidations: await annotationService.getAnnotationsByType('human')
}
```

## Label Studio Integration

Annotations automatically populate Label Studio tasks:
- `test.*` annotations provide expected outcomes
- `llm.*` annotations show AI predictions for validation
- `human.*` annotations preserve previous reviews
- Clean traces presented for unbiased analysis

## LLaMA-Factory Training

Prefix-based data stratification:
- **Gold tier**: Traces with `human.*` annotations
- **Silver tier**: Traces with `test.*` annotations
- **Bronze tier**: Traces with only `llm.*` annotations

### Phase 1: Core Infrastructure
- Create annotations table in ClickHouse
- Implement AnnotationService with Effect-TS
- Add validation for annotation targets
- Set up audit trail logging

### Phase 2: Integration
- Connect to Feature-005a capture sessions
- Integrate with feature flag controller
- Link to ADR-016 investigation flow
- Add UI for manual annotations

### Phase 3: Anti-Contamination
- Implement AnnotationSanitizer service
- Add environment-based validation rules
- Create training data export pipeline
- Set up contamination monitoring alerts

### Phase 4: Training Pipeline
- Export to Label Studio format
- Generate LLaMA-Factory datasets
- Implement data quality stratification
- Add validation metrics

### Data Quality
- **Zero contamination**: No test/train prefixes in production
- **Complete audit trail**: 100% of annotations tracked
- **Clean training data**: Validated no annotation leakage

### Operational
- **Query performance**: <100ms annotation lookups
- **Storage efficiency**: <5% overhead vs inline attributes
- **Annotation coverage**: >90% of test scenarios annotated

### Training Effectiveness
- **Label quality**: >95% human validation agreement
- **Data stratification**: Clear gold/silver/bronze tiers
- **Model improvement**: >30% accuracy gain with annotations

## Benefits

1. **Immutable Telemetry**: Core data never modified
2. **Rich Metadata**: Capture full diagnostic context
3. **Clean Training**: Models learn patterns, not labels
4. **Investigation History**: Complete diagnostic journey
5. **Flexible Queries**: Join annotations only when needed
6. **Production Safety**: Test annotations physically separated
7. **Scalable Architecture**: Independent optimization possible

## Security Considerations

- Separate permissions: Traces (read-only) vs annotations (read-write)
- No PII in annotations
- Complete audit logging
- Environment-based access control

## References

- Parent feature: Feature-005 (#103)
- OTLP Capture: Feature-005a (#110)
- Model Training: Feature-012 (#109)
- LLM Diagnostics: ADR-016 (#97)
